File: 4.8-bc/py/stage-3-hexagonal/app/notifications/app/service.py
"""Notifications use case — Send.

Called by the HTTP handler (POST /notifications/test) and by event
subscribers (welcome email on UserRegistered, etc).
"""
from __future__ import annotations

from datetime import UTC, datetime
from uuid import UUID, uuid4

from app.notifications.domain.notification import (
    Channel,
    Notification,
    Sender,
)
from app.notifications.domain.repository import Repository


class Service:
    def __init__(self, repo: Repository, sender: Sender) -> None:
        self._repo = repo
        self._sender = sender

    async def send(
        self, user_id: UUID, channel: Channel, payload: str
    ) -> Notification:
        n = Notification(
            id=uuid4(),
            user_id=user_id,
            channel=channel,
            payload=payload,
            sent_at=None,
        )
        await self._sender.send(n)
        n.sent_at = datetime.now(UTC)
        await self._repo.save(n)
        return n

File: 4.8-bc/go/stage-2-feature-first/app/notifications/service.go
// Package notifications — Notifications feature: відправка email/push.
package notifications

import (
	"context"
	"log"
	"time"

	"github.com/google/uuid"
)

type Service struct {
	repo *Repository
}

func NewService(repo *Repository) *Service {
	return &Service{repo: repo}
}

func (s *Service) Send(ctx context.Context, userID uuid.UUID, channel, payload string) (*Notification, error) {
	now := time.Now().UTC()
	n := Notification{
		ID:      uuid.New(),
		UserID:  userID,
		Channel: channel,
		Payload: payload,
		SentAt:  &now,
	}
	log.Printf("[stub-sender] %s -> user=%s payload=%q", channel, userID, payload)
	if err := s.repo.Save(ctx, n); err != nil {
		return nil, err
	}
	return &n, nil
}
